### PR TITLE
Add support for CUDA 12.1

### PR DESCRIPTION
**Pull request by Kerilk**

### CODE REVIEW

File: cuda/gen_babeltrace_cuda_model.rb [hunk 1]
```diff
@@ -31,9 +32,9 @@
   "int64_t" => 64,
   "uintptr_t" => 64,
   "size_t" => 64,
-  "CUtexObject" => 64,
-  "CUsurfObject" => 64,
-  "CUdeviceptr" => 64,
+  "CUtexObject_v1" => 64,
```
Comment (TApplencourt) [inline cuda/gen_babeltrace_cuda_model.rb:35]: Does that mean that old CUDA will not work anymore? I don't care just for my personal education :D

File: cuda/cuda_model.rb [hunk 2]
```diff
@@ -623,6 +652,49 @@ def initialize(command, name, size)
...
+class ArrayByRefMetaParameter < MetaParameter
+  attr_reader :size
+
+  def initialize(command, name, size)
+    @size = size
+    super(command, name)
+    a = command[name]
+    raise "Invalid parameter: #{name} for #{command.name}!" unless a
+    t = a.type
+    raise "Type is not a pointer: #{t}!" if !t.kind_of?(YAMLCAst::Pointer)
```
[the remark this answers is not in the document]
Comment (Kerilk) [reply cuda/cuda_model.rb]: Thanks, merging

File: cuda/cuda_model.rb [hunk 3]
```diff
@@ -40,15 +40,43 @@
...
+def transitive_closure_map(types, map)
+  sz = map.size
+  loop do
+    types.select { |t|
+      t.type.kind_of?(YAMLCAst::CustomType) && map.include?(t.type.name)
+    }.each { |t| map[t.name] = map[t.type.name] }
+    break if sz == map.size
+    sz = map.size
+  end
+end
```
[the remark this answers is not in the document]
Comment (Kerilk) [reply cuda/cuda_model.rb:63]: Wouldn't work as you would be selecting types which are already into the map.

File: cuda/Makefile.am [hunk 4]
```diff
@@ -266,19 +266,19 @@ nodist_libTracerCUDART_la_SOURCES = \
...
-libTracerCUDART_la_LDFLAGS += -version-number 11:0:3
+libTracerCUDART_la_LDFLAGS += -version-number 12:1:55
 libTracerCUDART_la_LIBADD = libcudarttracepoints.la
 
 install-exec-hook:
 	$(MKDIR_P) $(DESTDIR)$(pkglibdir)/cudart
 	$(MKDIR_P) $(DESTDIR)$(pkglibdir)/cuda
-	$(LN_S) -f $(DESTDIR)$(libdir)/libTracerCUDART.so.11.0.3 $(DESTDIR)$(pkglibdir)/cudart/libcudart.so.11.0
-	$(LN_S) -f $(DESTDIR)$(pkglibdir)/cudart/libcudart.so.11.0 $(DESTDIR)$(pkglibdir)/cudart/libcudart.so
+	$(LN_S) -f $(DESTDIR)$(libdir)/libTracerCUDART.so.12.1.55 $(DESTDIR)$(pkglibdir)/cudart/libcudart.so.12
```
[the remark this answers is not in the document]
Comment (Kerilk) [reply cuda/Makefile.am:275]: Yeah it feels strange but the libraries from Nvidia (and Cuda 12) are numbered this way so I decided to mirror it.

File: cuda/cuda_meta_parameters.yaml [hunk 5]
```diff
@@ -229,29 +294,72 @@ meta_parameters:
...
   cuMemGetAllocationPropertiesFromHandle:
     - [OutScalar, prop]
   cuMemRetainAllocationHandle:
     - [OutScalar, handle]
+  cuMemAllocAsync:
+    - [OutScalar, dptr]
+  cuMemAllocAsync_ptsz:
+    - [OutScalar, dptr]
+  cuMemPoolSetAttribute:
+    - [InLTTng, value, ctf_sequence_text, uint8_t, value_val, value, size_t, "attr == CU_MEMPOOL_ATTR_RELEASE_THRESHOLD ? sizeof(cuuint64_t) : attr == CU_MEMPOOL_ATTR_REUSE_FOLLOW_EVENT_DEPENDENCIES ? sizeof(int) : attr == CU_MEMPOOL_ATTR_REUSE_ALLOW_OPPORTUNISTIC ? sizeof(int) : attr == CU_MEMPOOL_ATTR_REUSE_ALLOW_INTERNAL_DEPENDENCIES ? sizeof(int) : attr == CU_MEMPOOL_ATTR_RESERVED_MEM_HIGH ? sizeof(cuuint64_t) : attr == CU_MEMPOOL_ATTR_USED_MEM_HIGH ? sizeof(cuuint64_t) : 0"]
```
[the remark this answers is not in the document]
Comment (Kerilk) [reply cuda/cuda_meta_parameters.yaml:330]: yeah I know...

File: cuda/cuda_model.rb [hunk 6]
```diff
@@ -40,15 +40,43 @@
...
+def transitive_closure(types, arr)
+  sz = arr.size
+  loop do
+    arr.concat( types.select { |t|
+      t.type.kind_of?(YAMLCAst::CustomType) && arr.include?(t.type.name)
+    }.collect { |t| t.name } ).uniq!
+    break if sz == arr.size
+    sz = arr.size
+  end
+end
```
[the remark this answers is not in the document]
Comment (Kerilk) [reply cuda/cuda_model.rb:52]: same issue as with the map, though using `filter_map` could arguably make the code more readable

File: cuda/gen_babeltrace_cuda_model.rb (hunk 1 above)
Comment (Kerilk) [reply cuda/gen_babeltrace_cuda_model.rb:35]: No this is just a transitivity issue, these types are now defined with the `_v1` type that didn't existed before. the versions without suffix still exist. This is the reason I had to add all those transitive closures, typdef on typedefs...

**Review by TApplencourt — APPROVED** [review]

**Pull request merged**